components/webgpu/canvas_context.rs

```diff
@@ -0,0 +1,771 @@
+    /// that will copy texture to this staging buffer.
+    ///
+    /// Caller must submit command buffer to queue.
+    fn prepare_load_texture_command_buffer(
+        &mut self,
+        texture_id: TextureId,
+        encoder_id: CommandEncoderId,
+        config: &ContextConfiguration,
+    ) -> Result<CommandBufferId, Box<dyn std::error::Error>> {
+        self.ensure_available(config)?;
+        let StagingBufferState::Available(buffer) = &self.state else {
+            unreachable!("Should be made available by `ensure_available`")
+        };
+        let device_id = buffer.device_id;
+        let command_descriptor = wgt::CommandEncoderDescriptor { label: None };
+        let (encoder_id, error) = self.global.device_create_command_encoder(
+            device_id,
+            &command_descriptor,
+            Some(encoder_id),
+        );
+        if let Some(error) = error {
+            return Err(error.into());
+        };
+        let buffer_info = wgt::TexelCopyBufferInfo {
+            buffer: self.buffer_id,
+            layout: wgt::TexelCopyBufferLayout {
+                offset: 0,
+                bytes_per_row: Some(config.stride()),
+                rows_per_image: None,
+            },
+        };
+        let texture_info = wgt::TexelCopyTextureInfo {
+            texture: texture_id,
+            mip_level: 0,
+            origin: wgt::Origin3d::ZERO,
+            aspect: wgt::TextureAspect::All,
+        };
+        let copy_size = wgt::Extent3d {
+            width: config.size.width,
+            height: config.size.height,
+            depth_or_array_layers: 1,
+        };
+        self.global.command_encoder_copy_texture_to_buffer(
+            encoder_id,
+            &texture_info,
+            &buffer_info,
+            &copy_size,
+        )?;
+        let (command_buffer_id, error) = self
+            .global
+            .command_encoder_finish(encoder_id, &wgt::CommandBufferDescriptor::default());
+        if let Some(error) = error {
+            return Err(error.into());
+        };
+        Ok(command_buffer_id)
+    }
+
+    /// Unmaps buffer or cancel mapping if in progress
+    fn unmap(&mut self) {
+        match self.state {
+            StagingBufferState::Unassigned | StagingBufferState::Available(_) => {},
+            StagingBufferState::Mapping(buffer) |
+            StagingBufferState::Mapped(MappedBuffer { buffer, .. }) => {
+                let _ = self.global.buffer_unmap(self.buffer_id);
+                self.state = StagingBufferState::Available(buffer)
+            },
+        }
+    }
+
+    /// Obtains snapshot from mapped buffer
+    fn snapshot(&self) -> Option<Snapshot> {
+        let StagingBufferState::Mapped(mapped) = &self.state else {
+            return None;
+        };
+        let format = match mapped.image_format {
+            ImageFormat::RGBA8 => SnapshotPixelFormat::RGBA,
+            ImageFormat::BGRA8 => SnapshotPixelFormat::BGRA,
+            _ => unimplemented!(),
```

Comment:
Better to return `None` here rather than crashing?

```suggestion
            _ => unimplemented!(),
```